File: packages/core/src/services/stream-processor/stream-processor.batch.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { StreamProcessorService } from './index';
import type { StreamUpdate, StreamUpdateReceiver } from './index';
import type { WasmProcessor, WasmStreamUpdate } from './wasm-types';

/**
 * Tests for `ingestMany` bulk insert on StreamProcessorService.
 *
 * A batched ingest (e.g. sync fetching N missing records) used to emit one
 * stream update per record, making the UI render row-by-row. ingestMany
 * collapses those into a single coalesced update per affected query.
 */

function makeLogger(): any {
  const noop = () => {};
  const logger: any = {
    debug: noop,
    info: noop,
    warn: noop,
    error: noop,
    trace: noop,
  };
  logger.child = () => logger;
  return logger;
}

/** Records every dispatched update so we can count notifications. */
class RecordingReceiver implements StreamUpdateReceiver {
  public received: StreamUpdate[] = [];
  onStreamUpdate(update: StreamUpdate): void {
    this.received.push(update);
  }
}

/**
 * Build a service with a stubbed WASM processor. The processor accumulates
 * ingested ids and returns the *full* materialized array on every call (as the
 * real WASM does), so coalescing's last-write-wins must yield the final array.
 */
function makeService(queryHashesFor: (id: string) => string[]) {
  const svc = new StreamProcessorService(
    {} as any,
    {} as any,
    { get: async () => undefined, set: async () => {} } as any,
    makeLogger()
  );

  const ingestedByQuery = new Map<string, Array<[string, number]>>();
  const mockProcessor: Partial<WasmProcessor> = {
    ingest: (_table, _op, id, record: any): WasmStreamUpdate[] => {
      const version = record?._00_rv ?? 1;
      const updates: WasmStreamUpdate[] = [];
      for (const queryHash of queryHashesFor(id)) {
        const arr = ingestedByQuery.get(queryHash) ?? [];
        arr.push([id, version]);
        ingestedByQuery.set(queryHash, arr);
        updates.push({ query_id: queryHash, result_data: [...arr] });
      }
      return updates;
    },
  };
  // processor is private and normally set by init() via WASM; inject the stub.
  (svc as any).processor = mockProcessor;
  return svc;
}

describe('StreamProcessor ingestMany bulk insert', () => {
  let receiver: RecordingReceiver;

  beforeEach(() => {
    receiver = new RecordingReceiver();
  });

  it('emits one update per record when ingesting one-by-one (baseline)', () => {
    const svc = makeService(() => ['q1']);
    svc.addReceiver(receiver);

    svc.ingest('user', 'CREATE', 'user:1', { id: 'user:1', _00_rv: 1 });
    svc.ingest('user', 'CREATE', 'user:2', { id: 'user:2', _00_rv: 1 });
    svc.ingest('user', 'CREATE', 'user:3', { id: 'user:3', _00_rv: 1 });

    expect(receiver.received).toHaveLength(3);
  });

  it('coalesces a bulk insert into a single update per query with the final array', () => {
    const svc = makeService(() => ['q1']);
    svc.addReceiver(receiver);

    svc.ingestMany([
      { table: 'user', op: 'CREATE', id: 'user:1', record: { id: 'user:1', _00_rv: 1 } },
      { table: 'user', op: 'CREATE', id: 'user:2', record: { id: 'user:2', _00_rv: 1 } },
      { table: 'user', op: 'CREATE', id: 'user:3', record: { id: 'user:3', _00_rv: 1 } },
    ]);

    // Nothing dispatched until the whole batch is ingested, then exactly one
    // coalesced update.
    expect(receiver.received).toHaveLength(1);
    const update = receiver.received[0];
    expect(update.queryHash).toBe('q1');
    // Last-write-wins carries the full materialized array (all 3 records).
    expect(update.localArray).toHaveLength(3);
    // Coalesced updates take DataModule's immediate (non-debounced) path.
    expect(update.op).toBe('CREATE');
    expect(typeof update.materializationTimeMs).toBe('number');
  });

  it('coalesces independently per affected query', () => {
    // user:1 -> q1, user:2 -> q1 & q2, user:3 -> q2
    const svc = makeService((id) =>
      id === 'user:1' ? ['q1'] : id === 'user:2' ? ['q1', 'q2'] : ['q2']
    );
    svc.addReceiver(receiver);

    svc.ingestMany([
      { table: 'user', op: 'CREATE', id: 'user:1', record: { id: 'user:1', _00_rv: 1 } },
      { table: 'user', op: 'CREATE', id: 'user:2', record: { id: 'user:2', _00_rv: 1 } },
      { table: 'user', op: 'CREATE', id: 'user:3', record: { id: 'user:3', _00_rv: 1 } },
    ]);

    expect(receiver.received).toHaveLength(2);
    const byHash = Object.fromEntries(receiver.received.map((u) => [u.queryHash, u]));
    expect(Object.keys(byHash).sort()).toEqual(['q1', 'q2']);
  });

  it('is a no-op for an empty batch and leaves the window closed', () => {
    const svc = makeService(() => ['q1']);
    svc.addReceiver(receiver);

    svc.ingestMany([]);

    expect(receiver.received).toHaveLength(0);
    // A subsequent single ingest dispatches normally (window never opened).
    svc.ingest('user', 'CREATE', 'user:1', { id: 'user:1', _00_rv: 1 });
    expect(receiver.received).toHaveLength(1);
  });
});

File: packages/spooky_core/test/stream_processor_batch_test.dart
import 'package:spooky_core/src/ffi/stream_update.dart';
import 'package:spooky_core/src/services/logger/logger.dart';
import 'package:spooky_core/src/services/persistence/memory_persistence.dart';
import 'package:spooky_core/src/services/stream_processor/stream_processor_service.dart';
import 'package:test/test.dart';

/// Tests for the batch-coalescing window added to [StreamProcessorService].
///
/// A batched ingest (e.g. sync fetching N missing records) used to emit one
/// stream update per record, making the UI render row-by-row. beginBatch/
/// endBatch collapse those into a single coalesced update per affected query.
void main() {
  final logger = SpookyLogger.root('test');

  group('StreamProcessor batch coalescing', () {
    late StreamProcessorService sp;
    late List<StreamUpdate> received;

    setUp(() async {
      sp = StreamProcessorService(MemoryPersistenceClient(), logger);
      await sp.init();
      sp.seedPermissionsFromSchema(
          'DEFINE TABLE user SCHEMAFULL PERMISSIONS FOR select WHERE true;');
      sp.registerQueryPlan(QueryPlanConfig(
        queryHash: 'q1',
        surql: 'SELECT * FROM user',
        params: {},
        ttl: '10m',
        lastActiveAt: DateTime.utc(2026),
      ));

      received = [];
      sp.addReceiver(_CapturingReceiver(received.add));
    });
    tearDown(() => sp.close());

    void ingestUser(int n) => sp
        .ingest('user', 'CREATE', 'user:$n', {'id': 'user:$n', 'name': 'u$n'});

    test('emits one update per record when NOT batching (baseline / the bug)',
        () {
      // This is the row-by-row behavior the user is seeing: three records, three
      // notifications, so the UI re-renders three times.
      ingestUser(1);
      ingestUser(2);
      ingestUser(3);

      expect(received, hasLength(3));
    });

    test('coalesces a batch into a single update with the final array', () {
      sp.beginBatch();
      try {
        ingestUser(1);
        ingestUser(2);
        ingestUser(3);
        // Nothing dispatched until endBatch.
        expect(received, isEmpty);
      } finally {
        sp.endBatch();
      }

      // Exactly one coalesced update, carrying the full materialized array.
      expect(received, hasLength(1));
      final update = received.single;
      expect(update.queryHash, 'q1');
      expect(update.localArray.map((e) => e.$1),
          containsAll(['user:1', 'user:2', 'user:3']));
      expect(update.localArray, hasLength(3));
      // Coalesced updates take DataModule's immediate (non-debounced) path.
      expect(update.op, 'CREATE');
      expect(update.materializationTimeMs, isNotNull);
    });

    test('batch-ingest of new records matching a FILTERED query emits them', () {
      // The app's initial down-sync path: register a record-filtered query, then
      // bulk-fetch the matching rows and ingest them as ONE batch (the live
      // per-record path doesn't fire for rows that pre-existed before the LIVE
      // subscription). The batch must materialize the matching rows into the
      // view — this is the path the live-driven e2e never exercises with NEW rows.
      sp.seedPermissionsFromSchema(
          'DEFINE TABLE game PERMISSIONS FOR select WHERE true;');
      sp.registerQueryPlan(QueryPlanConfig(
        queryHash: 'qg',
        surql: r'SELECT * FROM game WHERE database = $db ORDER BY sort_index ASC',
        // Param as the sanitized string form the app's local circuit receives
        // (a RecordId is stringified for the FFI's jsonEncode).
        params: {'db': 'game_database:DBS_x'},
        ttl: '10m',
        lastActiveAt: DateTime.utc(2026),
      ));
      received.clear();

      sp.beginBatch();
      try {
        for (var i = 0; i < 4; i++) {
          sp.ingest('game', 'CREATE', 'game:g$i', {
            'id': 'game:g$i',
            'database': 'game_database:DBS_x',
            'white': 'W',
            'black': 'B',
            'pgn': '1. e4 e5',
            'result': '1-0',
            'status': 'Finished',
            'sort_index': -i,
          });
        }
      } finally {
        sp.endBatch();
      }

      final gameUpdates = received.where((u) => u.queryHash == 'qg').toList();
      expect(gameUpdates, isNotEmpty,
          reason: 'filtered query must emit for a batch of matching rows');
      expect(gameUpdates.last.localArray.map((e) => e.$1),
          containsAll(['game:g0', 'game:g1', 'game:g2', 'game:g3']));
      expect(gameUpdates.last.localArray, hasLength(4));
    });

    test('endBatch is safe with no buffered updates; beginBatch is idempotent',
        () {
      sp.beginBatch();
      sp.beginBatch(); // no-op, must not reset/duplicate
      sp.endBatch();
      expect(received, isEmpty);

      // A subsequent non-batched ingest dispatches normally (window closed).
      ingestUser(1);
      expect(received, hasLength(1));
    });
  });
}

class _CapturingReceiver implements StreamUpdateReceiver {
  _CapturingReceiver(this._onUpdate);
  final void Function(StreamUpdate) _onUpdate;
  @override
  void onStreamUpdate(StreamUpdate update) => _onUpdate(update);
}
